File: Rust/rust_aoc/src/year_2015/d7_circuit.rs
use std::collections::HashMap;

pub fn part_one(lines: &Vec<String>) -> String {
    solve(lines).to_string()
}

pub fn part_two(lines: &Vec<String>) -> String {
    let part_one_result = part_one(lines);
    let mut copy = lines.clone();

    for line in copy.iter_mut() {
        if line.ends_with("-> b") {
            *line = part_one_result.clone() + " -> b";
            break;
        }
    }
    solve(&copy).to_string()
}

pub fn solve(lines: &Vec<String>) -> u16 {
    let mut resolved: HashMap<&str, u16> = HashMap::new();
    let entries: Vec<Entry> = lines.iter().map(|l| parse_entry(l)).collect();

    while resolved.len() < lines.len() {
        let to_add: Vec<(&str, u16)> = entries.iter()
            .filter(|e| !resolved.contains_key(e.output))
            .filter_map(|e| try_resolve(e, &resolved)).collect();
        
        if to_add.len() == 0 {
            panic!("Failed to find new entries");
        }
        for e in to_add {
            resolved.insert(e.0, e.1);
        }
    }

    *resolved.get(&"a").unwrap()
}
struct Entry<'a> {
    parts: Vec<&'a str>,
    output: &'a str
}

fn try_resolve<'a, 'b>(entry: &'a Entry, resolved: &HashMap<&'b str, u16>) -> Option<(&'a str, u16)> {
    match entry.parts.len() {
        1 => match try_resolve_part(entry.parts[0], resolved) {
                Some(v) => Some((&entry.output, v)),
                None => None
        },
        2 => match try_resolve_part(entry.parts[1], resolved) {
                Some(v) => Some((&entry.output, !v)),
                None => None
        },
        3 => {
            let lhs = try_resolve_part(entry.parts[0], resolved);
            let rhs = try_resolve_part(entry.parts[2], resolved);
            match (lhs, entry.parts[1], rhs) {
                (Some(left), "AND", Some(right)) => Some((&entry.output, left & right)),
                (Some(left), "OR", Some(right)) => Some((&entry.output, left | right)),
                (Some(left), "LSHIFT", Some(right)) => Some((&entry.output, left << right)),
                (Some(left), "RSHIFT", Some(right)) => Some((&entry.output, left >> right)),
                _ => None
            }
        }
        _ => None
    }
}

fn try_resolve_part(part: &str, resolved: &HashMap<&str, u16>) -> Option<u16> {
    match resolved.get(part) {
        Some(v) => Some(*v),
        None => {
            match part.parse::<u16>() {
                Ok(res) => Some(res),
                Err(..) => None
            }
        }
    }
}
/*
hv OR hu -> hw
19138 -> b
NOT ft -> fu
*/

fn parse_entry(line: &str) -> Entry {
    let s: Vec<&str> = line.split(" -> ").collect();
    Entry{parts: s[0].split(" ").collect(), output: s[1]}
}

File: Cpp/AdventOfCode/src/2015/d7_Circuit.cpp
#pragma once
#include "2015/d7_Circuit.h"

SOLUTION(2015, 7) {
    using Map = Constexpr::SmallMap<std::string, std::string>;

    constexpr void Resolve(const std::string& expr, Map& mapping);

    constexpr u16 GetValue(const std::string& str, Map& mapping) {
        u16 result;
        if (mapping.contains(str)) {
            Resolve(str, mapping);
            Constexpr::ParseNumber(mapping.at(str), result);
        }
        else {
            Constexpr::ParseNumber(str, result);
        }
        return result;
    }
    constexpr void Resolve(const std::string& key, Map& mapping) {
        auto expr = mapping.at(key);
        auto s = Constexpr::Split(expr, " ");
        if (s.size() == 1) {
            mapping[key] = Constexpr::ToString(GetValue(std::string(s[0]), mapping));
        }
        else if (s.size() == 2) {
            auto valStr = std::string(s[1]);
            auto val = GetValue(valStr, mapping);
            mapping[key] = Constexpr::ToString(static_cast<u16>(~val));
        }
        else {
            auto lhsStr = std::string(s[0]);
            auto op = std::string(s[1]);
            auto rhsStr = std::string(s[2]);

            auto lhs = GetValue(lhsStr, mapping);
            auto rhs = GetValue(rhsStr, mapping);

            if (op == "AND") mapping[key] = Constexpr::ToString(lhs & rhs);
            else if (op == "OR") mapping[key] = Constexpr::ToString(lhs | rhs);
            else if (op == "LSHIFT") mapping[key] = Constexpr::ToString(lhs << rhs);
            else if (op == "RSHIFT") mapping[key] = Constexpr::ToString(lhs >> rhs);
        }
    }

    constexpr Map ParseInput(const auto& lines) {
        Map result;
        for (const auto& line : lines) {
            auto s = Constexpr::Split(line, " -> ");
            result[std::string(s[1])] = std::string(s[0]);
        }
        
        return result;
    }

    constexpr std::string Solve(const auto& lines) {
        auto map = ParseInput(lines);
        Resolve("a", map);
        return map.at("a");
    }

    PART_ONE() {
        return Solve(Lines);
    }

    PART_TWO() {
        auto partOneResult = Solve(Lines);

        auto copy = CopyToVector(Lines);
        for (auto& line : copy) {
            if (line.substr(line.size() - 4) == "-> b") {
                line = partOneResult + " -> b";
                break;
            }
        }

        return Solve(copy);
    }

    TESTS() {
        std::vector<std::string> lines = {
            "123 -> x",
            "456 -> y",
            "x AND y -> d",
            "x OR y -> e",
            "x LSHIFT 2 -> f",
            "y RSHIFT 2 -> g",
            "NOT x -> h",
            "NOT y -> i"
        };

        auto map = ParseInput(lines);
        Resolve("d", map);
        Resolve("e", map);
        Resolve("f", map);
        Resolve("g", map);
        Resolve("h", map);
        Resolve("i", map);

        if (map.at("x") != "123") return false;
        if (map.at("y") != "456") return false;
        if (map.at("d") != "72") return false;
        if (map.at("e") != "507") return false;
        if (map.at("f") != "492") return false;
        if (map.at("g") != "114") return false;
        if (map.at("h") != "65412") return false;
        if (map.at("i") != "65079") return false;


        lines = {
            "a -> b",
            "b -> c",
            "c -> d",
            "42 -> a"
        };

        map = ParseInput(lines);
        Resolve("d", map);
        if (map.at("a") != "42") return false;
        if (map.at("b") != "42") return false;
        if (map.at("c") != "42") return false;
        if (map.at("d") != "42") return false;

        return true;
    }
}
